src/expressions.rst

```diff
@@ -312,9 +312,11 @@ A :t:`constant context` is a :t:`construct` that requires a
 * :dp:`fls_ucFupTeCyylb`
   The :t:`block expression` of a :t:`const block expression`.
 
-:dp:`fls_ox6sgl9n43g2`
-It is a static error to create a :t:`mutable reference` in a
-:t:`constant context`.
+* :dp:`fls_zyuxqty09SDO`
+  All forms of :t:`[borrow]s` except those of expressions that would be subject to
+  :t:`drop scope extension`,
+  and which are either :t:`[mutable borrow]s`
+  or borrows of expressions that result in values with :t:`interior mutability`.
```

Comment:
Two things.  One, it's OK for the expression to be subject to drop scope extension.  It's not OK for that extension to extend the scope outside of the initializer to the end of the program.

```rust
const _: () = { let x = { &mut 0u8 }; x; }; //~ OK
//                             ^^^
// The drop scope of this temporary is extended.
```

Two, the comma in "..., and which are either..." causes me to read it as a clause independent from "except...".  I.e., this says (applying commutativity) "all  forms of borrows 1) that are either mutable or result in values with interior mutability and 2) where the expression would not be subject to drop scope extension".

Reply:
This is correct, except that, if I'm reading the diff correctly, it's putting this language in the list of what constitutes a constant context rather than about what constitutes a constant expression.  We're defining constant expressions here, and the existing rules about "borrows" and "immutable borrow expressions" will need to be changed (and then that reflected in the changelog, etc.).